Make the domain tip dismissible

dzver opened the pull request:
Per @simison's request. There's no particular reason why this particular tip should not be dismissible other than maybe looking uglier with the x.

In case a person doesn't want to use a custom domain, they should be able to dismiss the ad to never see it again.

Note that this tip might be a good fit to be a JITM

<!--
Link a related issue to this PR. If the PR does not immediately resolve the issue,
for example, it requires a separate deployment to production, avoid
using the "fixes" keyword and instead attach the [Status] Fix Inbound label to
the linked issue.
-->

Related to #

## Proposed Changes

Before:
<img width="1175" alt="Screenshot 2023-07-24 at 16 06 13" src="https://github.com/Automattic/wp-calypso/assets/82778/c922e0ba-034b-4c49-b859-746bc79b26ae">

After:
<img width="1175" alt="Screenshot 2023-07-24 at 16 05 48" src="https://github.com/Automattic/wp-calypso/assets/82778/9b20a89a-8ff4-42ad-87b8-e7fffa6ca023">

## Testing Instructions

1. With the store sandbox enabled, purchase a plan but not a domain
2. Go to /stats/insights/

The upsell should be dismissible

Once the setting is persisted, it can be reset with wpsh, details in 31588-pb

<!--
Add as many details as possible to help others reproduce the issue and test the fix.
"Before / After" screenshots can also be very helpful when the change is visual.
-->

*

## Pre-merge Checklist

<!--
Complete applicable items on this checklist **before** merging into trunk. Inapplicable items can be left unchecked.

Both the PR author and reviewer are responsible for ensuring the checklist is completed.
-->

- [ ] Has the general commit checklist been followed? (PCYsg-hS-p2)
- [ ] [Have you written new tests](https://wpcalypso.wordpress.com/devdocs/docs/testing/index.md) for your changes?
- [ ] Have you tested the feature in Simple (P9HQHe-k8-p2), Atomic (P9HQHe-jW-p2), and self-hosted Jetpack sites (PCYsg-g6b-p2)?
- [ ] Have you checked for TypeScript, React or other console errors?
- [ ] Have you used memoizing on expensive computations? More info in [Memoizing with create-selector](https://github.com/Automattic/wp-calypso/blob/trunk/packages/state-utils/src/create-selector/README.md) and [Using memoizing selectors](https://react-redux.js.org/api/hooks#using-memoizing-selectors) and [Our Approach to Data](https://github.com/Automattic/wp-calypso/blob/trunk/docs/our-approach-to-data.md)
- [ ] Have we added the "[Status] String Freeze" label as soon as any new strings were ready for translation (p4TIVU-5Jq-p2)?
- [ ] For changes affecting Jetpack: Have we added the "[Status] Needs Privacy Updates" label if this pull request changes what data or activity we track or use (p4TIVU-ajp-p2)?

simison commented:
Any prop to make the banner still clickable? Either the whole banner, or leaving CTA there together with dismissal (which ever is a typical pattern in Calypso). Now I can't click the banner. :-)

dzver commented:
> Any prop to make the banner still clickable? Either the whole banner, or leaving CTA there together with dismissal (which ever is a typical pattern in Calypso). Now I can't click the banner. :-)

Wow, this is very unexpected. Thanks for catching it. I'll figure it out.

dzver commented:
The `DismissibleCard` is not accepting href and I'm not sure why, most likely because of UI/design reasons. I added it to the diff but the issue is that the whole banner is clickable and points to the domains with the exception of the X, which dismisses the link so a person can click the X and miss it, and go to /domains.

I added it but will need to go over all uses of dismissable banners to make sure nothing is broken.

chriskmnds commented on client/components/banner/index.jsx:
```diff
@@ -314,6 +314,7 @@ export class Banner extends Component {
 					preferenceName={ dismissPreferenceName }
 					temporary={ dismissTemporary }
 					onClick={ this.handleDismiss }
+					href={ ( disableHref || callToAction ) && ! forceHref ? null : this.getHref() }
```
This was a little hard for me to trace out. It's a quite complex condition, not sure if splitting it out in one or two variables might help (I'm guessing it will just be the same).

Just trying to spell it out in words to make sure I understand: "If href is disabled or there is a call to action, and we do not force Href on the card, then return null, otherwise call the function to compute the href"

chriskmnds reviewed (commented):
Tested and left a comment in the code (not sure if anything to do there though). Some observations:

- On testing, this is how it shows up for me, using the live container/image (in Chrome):

<img width="500" alt="Screenshot 2023-08-08 at 1 00 34 PM" src="https://github.com/Automattic/wp-calypso/assets/1705499/b13552e6-832e-483d-acf2-55d8289b8dff">

- Not sure if due to the overlapping icons, but I could not just dismiss the banner. It would always take me to the domains page anywhere I clicked.
- Dismissing the banner once with site "foo" and going back with site "bar" will still no longer show the banner. Is this expected?
- Also, on design, the icon (whether "x" or ">") seems to be immediately and directly connected in my mind to the banner being highlighted (so it's like I expect the "x" to take effect when clicking anywhere else in the banner). Maybe if we turn the background surrounding "x" to light gray or something, so hovering over it would be obvious that it's a separate action within the "overall action" of the banner. 🤷

dzver committed:
Make the domain tip dismissible

Per @simison's request. There's no particular reason why this particular
tip should not be dismissible other than maybe looking uglier with the
x.

In case a person doesn't want to use a custom domain, they should be
able to dismiss the ad to never see it again.

dzver committed:
Clicking the X shouldn't redirect to domains

chriskmnds reviewed (approved):
Tests out. I can close banner as expected when hitting the `x` icon. I can't think of any caveats with using `event.preventDefault` below, outside of breaking the `href` when present (which is what we want to achieve?) 🤔
